### PR TITLE
21536: Clap App!

### PR DESCRIPTION
**Pull request by Menithal**
### Content Update: No Client changes.
Your avatar can finally clap!

This Tablet app allows both HMD users with hand-controllers and keyboard users to clap!

Keyboard users can clap by simply pressing K.

HMD users will have to put their hands together and literally clap! However with this you can also control the amount of volume the clap is at, by simply clapping gently or hard. Just be careful to to smash your controllers together! Best way to do this is to angle your controllers slightly.

The app also adds a new Menu into Tools. This allows you to turn off clapping particles or turn on a helper mode for those who wish master clapping or simply want to venture into scripting! (available to HMD users only)

#HMD How to:
## User Helper mode:
You can find this setting under Clap > 
Note: The idea is not to smash controllers together, you will actually clap at a more leisurely distance.

- Hold your hands in front, keep them separate from each other.
- Rotate the controllers so that your palms will point towards each other, and the lines turn green.
- When Aligned both lines will be green: move your hands close enough until you see a yellow line appear between your hands, and your avatar's fingers will go straight.
- At this distance, you can register a clap. Moving at speed to this distance, or smaller will cause a clap.
- You can also rotate your hands while moving them closer to cause a clap.
- The faster you rotate your hand to a clap, or  faster you move to the area will also cause a clap

Doing the clapping motion at this distance or smaller will allow you to clap, but you have to turn your hands off alignment (one red) or get out of the distance to reset for a new clap. 

Practice will allow you do to golf claps or sarcastic claps..

## Known Issues:

- **BUG:** Same as the sign chat: When using Keyboard clap, When Avatar joint positions are set manually via `MyAvatar.setJointRotation`, Hands and fingers do not set joints and instead have an animation that cannot be overwritten!
- Clap angle behavior can be a bit tricky to get right
- App was developed on Vive, may need fine tuning for the Rift
- K is used instead of C to allow C to be used for crouch, communication among others in potential use cases.

# Test Plan
## Persistance Keyboard, and Basic UI appearance Test

1. Run clapApp script (either [through the build](https://github.com/Menithal/hifi/blob/21536/unpublishedScripts/marketplace/clap/clapApp.js) or [Marketplace Draft](http://mpassets.highfidelity.com/797180be-1d5d-4465-816e-fada92aa727a-v1/clapApp.js).
2. By default it will be turned on
3. You should see a menu on the top menu bar "Clap" with "Enable Clap Particles" and "Enable Clap Helper". These will be tested later
4. Press K. Your avatar should do a clap
5. Press K repeatedly, Each sound after each other should sound different from the last
5. Turn off Clap.
6. Press K. Your avatar should not clap.
7. "Clap" Menu should no longer appear.
8. Restart High Fidelity
9. Press K. Your avatar should not clap
10. Observe that the clap is off.
11. Turn Clap on.
12. "Clap" Menu should appear
13. Press K. Your avatar should clap. 
14. Goto "Clap" > "Enable Clap Particles" and turn it off
15. Press K Your avatar should clap. There should be no particles.

## Usability: This requires tests with both the Vive and the Rift

1. Get on HMD with Hand controllers.
2. Goto "Clap > Enable Clap Helper"
3. Two line debug helpers will appear on your hands. 
4. The line helpers will turn red when not facing another helper, with some tolerance.
5. Moving hands closer together while green will animate your hands to a "Flat" pose. a Yellow line will appear in between the hands. This is the distance at which, when this is done at velocity will allow you to clap.
6. Try Clapping
7. Turn off Helper
8. Try Clapping without helpers
9. Experiment and give feed back

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3235/HighFidelity-Beta-PR11344-40beb4bca8acbd1e6616a68f47fbe91cc57b04a1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3235/HighFidelity-Beta-PR11344-40beb4bca8acbd1e6616a68f47fbe91cc57b04a1.exe

File: unpublishedScripts/marketplace/clap/clapApp.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+// Load up engine
+var APP_NAME = "CLAP";
+var ClapEngine = Script.require(Script.resolvePath("scripts/ClapEngine.js?v9"));
+var tablet = Tablet.getTablet("com.highfidelity.interface.tablet.system");
+
+// Define Menu
+var blackIcon = Script.resolvePath("icons/tablet-icons/clap-a.svg?foxv2");
+var whiteIcon = Script.resolvePath("icons/tablet-icons/clap-i.svg?foxv2");
+
+if ( Settings.getValue("clapAppEnabled") === undefined){
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/clap/clapApp.js]: Styling: space before {

**Comment by Menithal** [comment]
@misslivingrose Updated js with the styling fix. forgot I didn't have auto style on save on... :(

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3273/HighFidelity-Beta-PR11344-59746c564b1386867471c4a110b70bbbb72a31d9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3273/HighFidelity-Beta-PR11344-59746c564b1386867471c4a110b70bbbb72a31d9.exe

**Pull request merged**